soroban-env-host/src/host/frame.rs

```diff
@@ -372,19 +372,32 @@ impl Host {
         res
     }
 
-    /// Returns [`Hash`] contract ID from the VM frame at the top of the context
-    /// stack, or a [`HostError`] if the context stack is empty or has a non-VM
-    /// frame at its top.
-    pub(crate) fn get_current_contract_id_opt_internal(&self) -> Result<Option<Hash>, HostError> {
-        self.with_current_frame(|frame| match frame {
+    fn opt_contract_id_from_frame(&self, frame: &Frame) -> Result<Option<Hash>, HostError> {
+        match frame {
             Frame::ContractVM { vm, .. } => Ok(Some(vm.contract_id.metered_clone(self)?)),
             Frame::HostFunction(_) => Ok(None),
             Frame::StellarAssetContract(id, ..) => Ok(Some(id.metered_clone(self)?)),
             #[cfg(any(test, feature = "testutils"))]
             Frame::TestContract(tc) => Ok(Some(tc.id.metered_clone(self)?)),
+        }
+    }
+
+    /// Internal diagnostic use only. Returns a `None` if current frame don't exist
+    /// (in the case of calling directly from a test).
+    pub(crate) fn get_current_contract_id_from_opt_frame(&self) -> Result<Option<Hash>, HostError> {
+        self.with_current_frame_opt(|opt_frame| match opt_frame {
+            Some(frame) => self.opt_contract_id_from_frame(frame),
+            None => Ok(None),
         })
     }
 
+    /// Returns [`Hash`] contract ID from the VM frame at the top of the context
+    /// stack, or a [`HostError`] if the context stack is empty or has a non-VM
+    /// frame at its top.
+    pub(crate) fn get_current_contract_id_opt_internal(&self) -> Result<Option<Hash>, HostError> {
```

Comment:
This comment isn't right in its modified form -- now it only returns an error if the context stack is empty, and returns `None` if there's a non-VM frame on top of stack.

In general I think this function and the newly introduced one are sufficiently similar and their names are sufficiently uninformative that it's unlikely callers will know which they want (considering they have the same signature). Could we look at the call sites, figure out which cases are actually important to differentiate, and make functions that are clear about which cases they return and why?

Reply:
Consolidated these two internal functions into one. 
I think we are using some of them in test workflow which needs to handle empty stack frame or `None` contract_id. 
@dmkozh may know better if we can clean it up further.

Reply:
I don't really know about all the use cases, but it seems like these can probably be consolidate into `Result<Hash, HostError>` (which should never error during normal execution, unless budget is exceeded). Tests can do `unwrap_or` or whatever is appropriate.

Reply:
There are two situations here that are "possibly erroneous, possibly things we want to tolerate (not return errors for)":

  1. Empty context stack. Currently `with_current_frame_opt` tolerates this and returns `Ok(None)` for empty context.
  2. Non-contract frame (`Frame::HostFunction(...)`). Currently `get_current_contract_id_opt_internal` tolerates this and returns `Ok(None)` for a non-contract frame.

I think it's reasonable that some callers will want to treat these contexts as tolerable (eg. to emit events before a contract is running, such as the "we're about to call this contract" event or whatever).

I think it's probably _not_ reasonable for the cases that want to tolerate "empty context" to fail to tolerate "non-contract frame" (or vice-versa). I think if a caller tolerates one, it probably wants to tolerate the other.

As evidence for this, I note that the only caller of `get_current_contract_id_opt_internal`  that passes `true` (i.e. indicating that it _wants_ to differentiate those cases) is `get_current_contract_id_internal`, and it immediately treats the `None` case as an error as well, i.e. it collapses the two cases it just asked to differentiate back into a single case: either no-context or non-contract-context both result in an error return from it.

I therefore propose:

  - `get_current_contract_id_opt_internal` gets loses its boolean argument
  - its `None` branch just becomes `None => Ok(None)`, i.e. it collapses both cases into `Ok(None)`

Reply:
(applied this fix myself)